service/history/queues/rescheduler.go

```diff
@@ -173,49 +176,68 @@ func (r *reschedulerImpl) reschedule() {
 	r.Lock()
 	defer r.Unlock()
 
-	r.metricsHandler.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.pq.Len()))
-
-	var failToSubmit []rescheduledExecuable
-	for !r.pq.IsEmpty() {
-		if r.timeSource.Now().Before(r.pq.Peek().rescheduleTime) {
-			break
-		}
-
-		rescheduled := r.pq.Remove()
-		executable := rescheduled.executable
-		if executable.State() == ctasks.TaskStateCancelled {
-			continue
-		}
-
-		submitted, err := r.scheduler.TrySubmit(executable)
-		if err != nil {
-			executable.Logger().Error("Failed to reschedule task", tag.Error(err))
+	r.metricsHandler.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.numExecutables))
+
+	now := r.timeSource.Now()
+	for key, pq := range r.pqMap {
```

Comment:
this means we reschedule all tasks for a given namespace until their channel full before we move on to next namespace, is that expected?

Reply:
Yeah, I am not very concern about that as 
1. submitting is should be much faster than the actual tasks processing. 
2. this is an issue when there're larger number of tasks needs to be rescheduled in one shard but the existing task channel is almost empty (max size 1k shared by all shards). So I think it's very rare, unless someone is doing shard targeting.

With that said, I think it may be worth to add a metrics for rescheduling_latency so that we can better understand this and we can implement more complex logic if this latency goes up. What do you think?